File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int ans = 0;

void dfs(vector<int> v, int target, int sum, int idx){
    if(idx == v.size()){
        if(sum == target)
            ans++;
        return;
    }
    dfs(v, target, sum + v[idx], idx + 1);
    dfs(v, target, sum - v[idx], idx + 1);
}


int solution(vector<int> numbers, int target) {
    dfs(numbers, target, 0, 0);
    return ans;
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    public int[] arr;
    public int cnt;
    
    public int solution(int[] numbers, int target) {
        cnt = 0;
        arr = numbers.clone();
        dfs(0, 0, target);    
        return cnt;
    }
    
    public void dfs(int idx, int sum, int target) {
        if(idx == arr.length) {
            if(sum == target)
                cnt++;
            return;
        }
        dfs(idx + 1, sum + arr[idx], target);
        dfs(idx + 1, sum - arr[idx], target);
    }
    
}